lib/exercises.js

```diff
@@ -1,33 +1,145 @@
 // This method will return an array of arrays.
 // Each subarray will have strings which are anagrams of each other
-// Time Complexity: ?
-// Space Complexity: ?
+// Time Complexity: O(n) n rep num of string m is length O(n mlog m)
+// Space Complexity: O(n)
+
+//   Output:
+// [
+//   ["ate", "eat", "tea"],
+//   ["nat", "tan"],
+//   ["bat"]
+// ]
 function grouped_anagrams(strings) {
-  throw new Error("Method hasn't been implemented yet!");
+  let results = {};
+
+  for (string of strings) {
+    sorted = string.split("").sort().join("");
+    if (results[sorted]) {
+      results[sorted].push(string);
+    } else results[sorted] = [string];
+  }
+
+  return Object.values(results);
+
+  //TODO:
+  // // group letters
+  // //each letter same count
+  // let letters = {};
+  // let results = {};
+
+  // for (let string of strings) {
+  //   currentCount = {};
+  //   for (let char of string) {
+  //     if (currentCount[char] === undefined) {
+  //       currentCount[char] = 1;
+  //     } else {
+  //       currentCount[char] += 1;
+  //     }
+  //   }
+  //   letters[string] = currentCount;
+  //   for (let string of strings) {
+  //     for (let char = ; )
+  //   }
+  // }
+
+  // console.log(letters[]);
 }
+// let strings = ["eat", "tea", "tan", "ate", "nat", "bat"];
+// console.log(grouped_anagrams(strings));
 
 // This method will return the k most common elements
 // in the case of a tie it will select the first occuring element.
-// Time Complexity: ?
-// Space Complexity: ?
+// Time Complexity: O(n log n)  where n reps number of unqine elements
+// Space Complexity: O(n) where n reps number of unqine elements
+// let list = [1, 1, 1, 2, 2, 3], k = 2
+// Output: [1, 2]
 function top_k_frequent_elements(list, k) {
-  throw new Error("Method hasn't been implemented yet!");
-}
+  // get counts
+  //compare counts
+  //return as array
+  let counts = {};
+  for (let num of list) {
+    // O(n) where n is length of list
+    if (counts[num]) {
+      counts[num] += 1;
+    } else {
+      counts[num] = 1;
+    }
+  }
+
+  sorted = Object.keys(counts).sort((a, b) => counts[b] - counts[a]); //O(m log m) where m is num unique elements
 
+  sorted = sorted.map((i) => parseInt(i));
 
+  return sorted.slice(0, k);
+}
+
+let list = [1, 1, 1, 2, 2, 3];
+let k = 2;
+top_k_frequent_elements(list, k);
 // This method will return true if the table is still
 //   a valid sudoku table.
 //   Each element can either be a ".", or a digit 1-9
-//   The same digit cannot appear twice or more in the same 
+//   The same digit cannot appear twice or more in the same
 //   row, column or 3x3 subgrid
-// Time Complexity: ?
+// Time Complexity: 0(n ^ 2) n reps number of rows
 // Space Complexity: ?
 function valid_sudoku(table) {
```

Comment:
This doesn't quite pass all the tests, but it looks like you're on a path toward a solution.